inc/msp/ByteVector.hpp

```diff
@@ -0,0 +1,411 @@
+     * remaining bytes will be consumed.
+     * @return True on successful unpack
+     */
+    bool unpack(Value<ByteVector>& val, size_t count = std::numeric_limits<size_t>::max()) {
+        return val.set() = unpack(val(),count); 
+    }
+    
+    /**
+     * @brief Unpacks scaled Value types (e.g. scaled int to floating point) as val = (packed_val/scale)-offset
+     * @tparam encoding_T data type used to store the scaled Value (usually an integral type)
+     * @tparam T1 type of output Value (usually a floating point type)
+     * @tparam T2 type of scale and offset coefficients
+     * @param val Destination of unpack operation
+     * @param scale Value of scaling to apply to the offset value
+     * @param offset Value of offset to apply to the input value (optional, defaults to 0)
+     * @return True if successful
+     */
+    template< typename encoding_T, typename T1, typename T2, 
+                typename std::enable_if<std::is_arithmetic<T1>::value, T1>::type* = nullptr,
+                typename std::enable_if<std::is_arithmetic<T2>::value, T2>::type* = nullptr >
+    bool unpack(Value<T1>& val, T2 scale = 1, T2 offset = 0) {
+        return val.set() = unpack<encoding_T>(val(),scale,offset);
+    }
+    
+    /**
+     * @brief Gives the number of bytes which have already been consumed by 
+     * unpack operations.
+     * @returns Number of bytes already consumed
+     */
+    std::size_t unpacking_offset() {
+        return offset;
+    }
+    
+    /**
+     * @brief Gives an iterator to the next element ready for unpacking
+     * @returns iterator to the next byte for unpacking
+     */
+    std::vector<uint8_t>::iterator unpacking_iterator() {
+        return this->begin()+offset;
+    }
+    
+    /**
+     * @brief Manually consumes data, thus skipping the values.
+     * @param count Number of bytes to consume
+     * @returns True if successful
+     * @returns False if there were not enough bytes to satisfy the request
+     */
+    bool consume(std::size_t count) {
+        if (count < -offset) return false;
+        if (count > this->size() - offset) return false;
+        offset += count;
+        return true;
+    }
+    
+    /**
+     * @brief Returns the number of bytes still avialable for unpacking
+     * @returns Number of bytes remaining
+     */
+    std::size_t unpacking_remaining() {
+        return this->size() - offset;
+    }
+    
+
+protected:
+    std::size_t offset;
+
+};
+
+
+
+/**
+ * @brief Definition of a pure virtual class used to indicate that a child
+ * class can pack itself into a ByteVector and unpack itself from a ByteVector.
+ */
+struct Packable {
+    virtual bool pack_into(ByteVector &data) const = 0;
+    virtual bool unpack_from(ByteVector &data) = 0;
+};
+
+typedef std::shared_ptr<ByteVector> ByteVectorPtr;
+typedef std::unique_ptr<ByteVector> ByteVectorUptr;
+/////////////////////////////////////////////////////////////////////
+/// Generic message types
```

Comment:
Is this comment (Generic message types) supposed to be here? It looks misplaced since there are no generic types in this file.